src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Services/BookmarkResolver.cs

```diff
@@ -0,0 +1,548 @@
+        }
+        catch (Exception ex)
+        {
+            Logger.LogError("Failed to classify", ex);
+            var result = Classification.Unknown(input ?? string.Empty);
+            return (false, result);
+        }
+    }
+
+    public Classification ClassifyOrUnknown(string input)
+    {
+        return TryClassify(input, out var c) ? c : Classification.Unknown(input);
+    }
+
+    private bool TryClassify(string? input, out Classification result)
+    {
+        try
+        {
+            bool success;
+
+            if (string.IsNullOrWhiteSpace(input))
+            {
+                result = Classification.Unknown(input ?? string.Empty);
+                success = false;
+            }
+            else
+            {
+                input = input.Trim();
+
+                // is placeholder?
+                var isPlaceholder = _placeholderParser.ParsePlaceholders(input, out var inputUntilFirstPlaceholder, out _);
+                success = ClassifyCore(input, out result, isPlaceholder, inputUntilFirstPlaceholder, _placeholderParser);
+            }
+
+            return success;
+        }
+        catch (Exception ex)
+        {
+            Logger.LogError($"Failed to classify bookmark \"{input}\"", ex);
+            result = Classification.Unknown(input ?? string.Empty);
+            return false;
+        }
+    }
+
+    private static bool ClassifyCore(string input, out Classification result, bool isPlaceholder, string inputUntilFirstPlaceholder, IPlaceholderParser placeholderParser)
+    {
+        // 1) Try URI parsing first (accepts custom schemes, e.g., shell:, ms-settings:)
+        //    File URIs must start with "file:" to avoid confusion with local paths - which are handled below, in more sophisticated ways -
+        //    as TryCreate would automatically add "file://" to bare paths like "C:\path\to\file.txt" which we don't want.
+        if (Uri.TryCreate(input, UriKind.Absolute, out var uri)
+            && !string.IsNullOrWhiteSpace(uri.Scheme)
+            && (uri.Scheme != Uri.UriSchemeFile || input.StartsWith("file:", StringComparison.OrdinalIgnoreCase))
+            && uri.Scheme != UriSchemeShell)
+        {
+            // http/https → Url; any other scheme → Protocol (mailto:, ms-settings:, slack://, etc.)
+            var isWeb = uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps;
+
+            result = new Classification(
+                isWeb ? CommandKind.WebUrl : CommandKind.Protocol,
+                input,
+                input,
+                string.Empty,
+                LaunchMethod.ShellExecute, // Shell picks the right handler
+                null,
+                isPlaceholder);
+
+            return true;
+        }
+
+        // 1a) We're a placeholder and start look like a protocol scheme (e.g. "myapp:{{placeholder}}")
+        if (isPlaceholder && !string.IsNullOrWhiteSpace(inputUntilFirstPlaceholder) && inputUntilFirstPlaceholder.Contains(':'))
```

Comment:
would 

```
this isn't a protocol: it's {garbage}
```

fall into this case, and try to treat `this isn't a protocol` as the scheme?

(not blocking, but good follow up)